modules/collect/src/main/java/com/opengamma/strata/collect/io/XmlFile.java

```diff
@@ -0,0 +1,282 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ * 
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.collect.io;
+
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.ByteSource;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.Unchecked;
+
+/**
+ * An XML file.
+ * <p>
+ * Represents an XML file together with the ability to parse it from a {@link ByteSource}.
+ * <p>
+ * This uses the standard StAX API to parse the file.
+ * Once parsed, the XML is represented as a DOM-like structure, see {@link XmlElement}.
+ * This approach is suitable for XML files where the size of the parsed XML file is
+ * known to be manageable in memory.
+ * <p>
+ * Note that the {@link XmlElement} representation does not express all XML features.
+ * No support is provided for processing instructions, comments or mixed content.
+ * In addition, it is not possible to determine the difference between empty content and no children.
+ * <p>
+ * There is only minimal support for namespaces.
+ * Information is retained in the format of the original XML document.
+ * Thus, elements and attributes names may have prefixes.
+ */
+public final class XmlFile {
+
+  /**
+   * The root element.
+   */
+  private final XmlElement root;
+  /**
+   * The map of references.
+   */
+  private final ImmutableMap<String, XmlElement> refs;
+
+  //-----------------------------------------------------------------------
+  /**
+   * Parses the specified source as an XML file to an in-memory DOM-like structure.
+   * <p>
+   * This parses the specified byte source expecting an XML file format.
+   * The resulting instance can be queried for the root element.
+   * 
+   * @param source  the XML source data
+   * @return the parsed file
+   * @throws UncheckedIOException if an IO exception occurs
+   * @throws IllegalArgumentException if the file cannot be parsed
+   */
+  public static XmlFile of(ByteSource source) {
+    return of(source, "");
+  }
+
+  /**
+   * Parses the specified source as an XML file to an in-memory DOM-like structure.
+   * <p>
+   * This parses the specified byte source expecting an XML file format.
+   * The resulting instance can be queried for the root element.
+   * <p>
+   * This supports capturing attribute references, such as an id/href pair.
```

Comment:
I'm not sure from the Javadoc what you mean by "capturing attribute references". More explanation or an example would be helpful.